core/java/com/google/instrumentation/stats/MeasurementDescriptor.java

```diff
@@ -0,0 +1,148 @@
+ */
+public final class MeasurementDescriptor {
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  /**
+   * Name of measurement, e.g. rpc_latency, cpu. Must be unique.
+   */
+  public final String name;
+
+  /**
+   * Detailed description of the measurement, used in documentation.
+   */
+  public final String description;
+
+  /**
+   * The units in which {@link MeasurementDescriptor} values are measured.
+   */
+  public final MeasurementUnit unit;
+
+  /**
+   * Constructor.
+   */
+  public MeasurementDescriptor(String name, String description, MeasurementUnit unit) {
+    this.name = StringUtil.sanitize(name);
+    this.description = description;
+    this.unit = unit;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return (obj instanceof MeasurementDescriptor)
+        && name.equals(((MeasurementDescriptor) obj).name);
+  }
+
+  @Override
+  public int hashCode() {
+    return name.hashCode();
+  }
+
+  @Override
+  public String toString() {
+    return name;
+  }
+
+  /**
+   * Fundamental units of measurement.
+   */
+  public enum BasicUnit {
+    SCALAR,
+    BITS,
+    BYTES,
+    SECS,
+    CORES,
+    MAX_UNITS;
+  }
+
+  /**
+   * MeasurementUnit lets you build compound units of the form
+   * 10^n * (A * B * ...) / (X * Y * ...),
+   * where the elements in the numerator and denominator are all BasicUnits.  A
+   * MeasurementUnit must have at least one BasicUnit in its numerator.
+   *
+   * <p>To specify multiplication in the numerator or denominator, simply specify
+   * multiple numerator or denominator fields.  For example:
+   *
+   * <p>- byte-seconds (i.e. bytes * seconds):
+   *     numerator: BYTES
+   *     numerator: SECS
+   *
+   * <p>- events/sec^2 (i.e. rate of change of events/sec):
+   *     numerator: SCALAR
+   *     denominator: SECS
+   *     denominator: SECS
+   *
+   * <p>To specify multiples (in power of 10) of units, specify a non-zero power10
+   * value, for example:
+   *
+   * <p>- MB/s (i.e. megabytes / s):
+   *     power10: 6
+   *     numerator: BYTES
+   *     denominator: SECS
+   *
+   * <p>- nanoseconds
+   *     power10: -9
+   *     numerator: SECS
+   */
+  public static final class MeasurementUnit {
+    /**
+     * Unit multiplier (i.e. 10^power10).
+     */
+    public final int power10;
+
+    /**
+     * Unit Numerators.
+     *
+     * <p>Note: This list is unmodifiable, attempts to update it will throw an
+     * UnsupportedOperationException.
+     */
+    public final List<BasicUnit> numerators;
+
+    /**
+     * Unit Denominators.
+     *
+     * <p>Note: This list is unmodifiable, attempts to update it will throw an
+     * UnsupportedOperationException.
+     */
+    public final List<BasicUnit> denominators;
+
+    /**
+     * Basic constructor that takes all fields.
+     */
+    public MeasurementUnit(int power10, List<BasicUnit> numerators, List<BasicUnit> denominators) {
```

Comment:
constructors like these are fine to start with, but they become laborious when you add fields later. factory and/or builder, or even auto-value are preferred imho